src/lib.rs

```diff
@@ -1,7 +1,256 @@
+    // created by SecretKey::parse
+    let result =
+        clear_stack_on_return::<_, Result<(Signature, RecoveryId), secp256k1::Error>>(1, || {
+            let secret_key = SecretKey::parse(key)?;
+            Ok(sign(&message, &secret_key))
+        });
+
+    // Shenanigans to satisfy the compiler.
+    // `?` doesn't work because returning impl Error
+    let (sig, recovery) = match result {
+        Ok(s) => s,
+        Err(e) => return Err(e),
+    };
+
+    Ok((sig.serialize(), recovery.serialize() + 27))
+}
+
 #[cfg(test)]
 mod tests {
+    use super::*;
+    use hex;
+    use std::convert::TryInto;
+
+    // No salt
+    struct DomainStruct {
+        name: String,
+        version: String,
+        chain_id: U256,
+        verifying_contract: Address,
+    }
+
+    impl StructType for DomainStruct {
+        const TYPE_NAME: &'static str = "EIP712Domain";
+        fn visit_members<T: MemberVisitor>(&self, visitor: &mut T) {
+            visitor.visit("name", &self.name);
+            visitor.visit("version", &self.version);
+            visitor.visit("chainId", &self.chain_id);
+            visitor.visit("verifyingContract", &self.verifying_contract);
+        }
+    }
+
+    struct Person {
+        name: String,
+        wallet: Address,
+    }
+    impl StructType for Person {
+        const TYPE_NAME: &'static str = "Person";
+        fn visit_members<T: MemberVisitor>(&self, visitor: &mut T) {
+            visitor.visit("name", &self.name);
+            visitor.visit("wallet", &self.wallet);
+        }
+    }
+
+    struct Mail {
+        from: Person,
+        to: Person,
+        contents: String,
+    }
+    impl StructType for Mail {
+        const TYPE_NAME: &'static str = "Mail";
+        fn visit_members<T: MemberVisitor>(&self, visitor: &mut T) {
+            visitor.visit("from", &self.from);
+            visitor.visit("to", &self.to);
+            visitor.visit("contents", &self.contents);
+        }
+    }
+
     #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+    fn spec_case() {
+        // Taken from the JSON RPC section of the spec,
+        // as well as the accompanying example here:
+        // https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.js
+
+        let domain = DomainStruct {
+            name: "Ether Mail".to_owned(),
+            version: "1".to_owned(),
+            chain_id: U256([
+                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                0, 0, 0, 1,
+            ]),
+            verifying_contract: Address(
+                (&(hex::decode("CcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC").unwrap())[..])
+                    .try_into()
+                    .unwrap(),
+            ),
+        };
+        let domain_separator = DomainSeparator::new(&domain);
+
+        let message = Mail {
+            from: Person {
+                name: "Cow".to_owned(),
```

Comment:
The spec is unclear. All it says is...

> The dynamic values bytes and string are encoded as a keccak256 hash of their contents.

It seems reasonable to assume utf-8 since that is what is used by Ethereum ABIs.